tls/s2n_crl.c

```diff
@@ -0,0 +1,85 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include "s2n_crl.h"
+#include "tls/s2n_connection.h"
+
+struct s2n_crl *s2n_crl_new(void) {
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_crl)));
+    PTR_GUARD_POSIX(s2n_blob_zero(&mem));
+
+    struct s2n_crl *crl = (struct s2n_crl *)(void*) mem.data;
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return crl;
+}
+
+int s2n_crl_load_pem(struct s2n_crl *crl, uint8_t *pem, size_t len) {
+    POSIX_ENSURE_REF(crl);
+    POSIX_ENSURE(crl->crl == NULL, S2N_ERR_INVALID_ARGUMENT);
+
+    struct s2n_blob pem_blob = { 0 };
+    POSIX_GUARD(s2n_blob_init(&pem_blob, pem, len));
+
+    struct s2n_stuffer pem_stuffer = { 0 };
+    POSIX_GUARD(s2n_stuffer_init(&pem_stuffer, &pem_blob));
+    POSIX_GUARD(s2n_stuffer_skip_write(&pem_stuffer, pem_blob.size));
+
+    DEFER_CLEANUP(struct s2n_stuffer der_out_stuffer = {0}, s2n_stuffer_free);
+    POSIX_GUARD(s2n_stuffer_growable_alloc(&der_out_stuffer, 2048));
+    POSIX_GUARD(s2n_stuffer_crl_from_pem(&pem_stuffer, &der_out_stuffer));
+
+    const uint8_t *data = der_out_stuffer.blob.data;
+    crl->crl = d2i_X509_CRL(NULL, &data, der_out_stuffer.blob.size);
```

Comment:
Can't the stuffer blob be larger than the data actually written to it? I don't think der_out_stuffer.blob.size is what you want here. I think you want s2n_stuffer_data_available(&der_out_stuffer).

How about:
```suggestion

    uint32_t data_size = s2n_stuffer_data_available(&der_out_stuffer);
    const uint8_t *data = s2n_stuffer_raw_read(&der_out_stuffer, data_size);
    crl->crl = d2i_X509_CRL(NULL, &data, der_out_stuffer.blob.size);
```

Reply:
I added this, thanks!